File: app/src/main/java/com/singword/app/data/local/wordbook/WordbookRepository.kt
package com.singword.app.data.local.wordbook

import android.content.Context
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import java.io.FileNotFoundException

class WordbookRepository(private val context: Context) {

    private val cache = mutableMapOf<String, Map<String, WordEntry>>()

    fun loadWordbook(meta: WordbookMeta): WordbookLoadResult {
        cache[meta.assetName]?.let { cached ->
            return WordbookLoadResult.Success(
                cached.mapValues { (_, entry) -> entry to meta.label }
            )
        }

        val assetPath = "wordbooks/${meta.assetName}.json"
        val json = try {
            context.assets.open(assetPath).bufferedReader().use { it.readText() }
        } catch (e: Exception) {
            if (e is FileNotFoundException || e.message?.contains("No such file", ignoreCase = true) == true) {
                return WordbookLoadResult.MissingAsset(assetPath)
            }
            return WordbookLoadResult.ParseError(assetPath, e.message ?: "无法读取词表")
        }

        return try {
            val type = object : TypeToken<List<WordEntry>>() {}.type
            val words: List<WordEntry> = Gson().fromJson(json, type)
            val map = words.associateBy { it.word.lowercase() }
            cache[meta.assetName] = map
            WordbookLoadResult.Success(map.mapValues { (_, entry) -> entry to meta.label })
        } catch (e: Exception) {
            WordbookLoadResult.ParseError(assetPath, e.message ?: "词表解析失败")
        }
    }

    fun loadEnabledWordbooks(enabled: List<WordbookId>): WordbookLoadResult {
        val result = mutableMapOf<String, Pair<WordEntry, String>>()
        for (id in enabled) {
            when (val loadResult = loadWordbook(WordbookCatalog.metaFor(id))) {
                is WordbookLoadResult.Success -> {
                    for ((word, pair) in loadResult.words) {
                        if (word !in result) {
                            result[word] = pair
                        }
                    }
                }
                is WordbookLoadResult.MissingAsset -> return loadResult
                is WordbookLoadResult.ParseError -> return loadResult
            }
        }
        return WordbookLoadResult.Success(result)
    }
}

File: ios/singword/Data/Wordbooks/WordbookRepository.swift
import Foundation

actor WordbookRepository {
    private var cache: [String: [String: WordEntry]] = [:]

    func loadWordbook(meta: WordbookMeta) -> WordbookLoadResult {
        if let cached = cache[meta.assetName] {
            return .success(words: cached.mapValues { entry in (entry, meta.label) })
        }

        let assetPath = "wordbooks/\(meta.assetName).json"
        guard let url = resolveWordbookURL(named: meta.assetName) else {
            return .missingAsset(assetPath: assetPath)
        }

        do {
            let data = try Data(contentsOf: url)
            let words = try JSONDecoder().decode([WordEntry].self, from: data)
            let mapped = words.reduce(into: [String: WordEntry]()) { result, entry in
                result[entry.word.lowercased()] = entry
            }
            cache[meta.assetName] = mapped
            return .success(words: mapped.mapValues { entry in (entry, meta.label) })
        } catch {
            return .parseError(assetPath: assetPath, reason: error.localizedDescription)
        }
    }

    func loadEnabledWordbooks(enabled: [WordbookId]) -> WordbookLoadResult {
        var result: [String: (WordEntry, String)] = [:]

        for id in enabled {
            let meta = WordbookCatalog.meta(for: id)
            switch loadWordbook(meta: meta) {
            case .success(let words):
                for (word, pair) in words where result[word] == nil {
                    result[word] = pair
                }
            case .missingAsset(let assetPath):
                return .missingAsset(assetPath: assetPath)
            case .parseError(let assetPath, let reason):
                return .parseError(assetPath: assetPath, reason: reason)
            }
        }

        return .success(words: result)
    }

    private func resolveWordbookURL(named assetName: String) -> URL? {
        if let url = Bundle.main.url(forResource: assetName, withExtension: "json", subdirectory: "wordbooks") {
            return url
        }
        if let url = Bundle.main.url(forResource: assetName, withExtension: "json", subdirectory: "Resources/wordbooks") {
            return url
        }
        return Bundle.main.url(forResource: assetName, withExtension: "json")
    }
}
